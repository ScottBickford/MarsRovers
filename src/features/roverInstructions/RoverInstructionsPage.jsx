import React, { useState } from 'react';
import { Grid, Label } from 'semantic-ui-react';
import RoverInstructionsInput from './RoverInstructionsInput';
import { RoverInstructionsOutput } from './RoverInstructionsOutput';
import { processRoverInstructions } from './roverMovements';
import { RoversCoverageSection } from '../roverCoverage/RoversCoverageSection';

const RoverInstructionsPage = () => {
  const [instructions, setInstructions] = useState('');
  const [roverMovements, setRoverMovements] = useState(null);
  const [loaded, setLoaded] = useState(false);
  const [error, setError] = useState(null);

  const handleSubmitInstructions = event => {
    event.preventDefault();
    setError(null);
    try {
      var movements = processRoverInstructions(instructions);
      setRoverMovements(movements);
    } catch (error) {
      setError(error.message || error);
    } finally {
      setLoaded(true);
    }
  };

  return (
    <div>
      <Grid columns={2}>
        <Grid.Column>
          <RoverInstructionsInput
            instructions={instructions}
            setInstructions={setInstructions}
            handleSubmitInstructions={handleSubmitInstructions}
          />
        </Grid.Column>
        <Grid.Column>
          {loaded && error && (
            <Label color='red' basic pointing='left'>
              {error}
            </Label>
          )}
          {loaded && !error && (
            <RoverInstructionsOutput roverMovements={roverMovements} />
          )}
        </Grid.Column>
      </Grid>
      {loaded && !error && (
        <RoversCoverageSection roverMovements={roverMovements} />
      )}
    </div>
  );
};

export default RoverInstructionsPage;
