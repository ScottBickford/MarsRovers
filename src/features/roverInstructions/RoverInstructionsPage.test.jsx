import React from 'react';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import RoverInstructionsPage from './RoverInstructionsPage';

describe('RoverInstructionsPage', () => {
  test('snapshot renders', () => {
    const component = renderer.create(<RoverInstructionsPage />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('passes all props to Input', () => {
    const { getByText } = render(<RoverInstructionsPage />);
    const buttonElement = getByText(/Submit/i);
    expect(buttonElement).toBeInTheDocument();
  });
});