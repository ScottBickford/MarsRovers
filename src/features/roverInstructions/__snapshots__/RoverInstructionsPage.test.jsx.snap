// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoverInstructionsPage snapshot renders 1`] = `
<div>
  <div
    className="ui two column grid"
  >
    <div
      className="column"
    >
      <form
        onSubmit={[Function]}
      >
        <div
          className="ui segments"
        >
          <div
            className="ui teal inverted segment"
          >
            <div
              className="ui header"
            >
              Input
            </div>
          </div>
          <div
            className="ui attached segment"
          >
            <textarea
              onChange={[Function]}
              placeholder="Enter your instructions here..."
              rows="10"
              style={
                Object {
                  "width": "100%",
                }
              }
              value=""
            />
            <div
              className="ui divider"
            />
            <button
              className="ui button"
              onClick={[Function]}
            >
              Submit
            </button>
          </div>
        </div>
      </form>
    </div>
    <div
      className="column"
    />
  </div>
</div>
`;
